hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/LogFileIterator.scala

```diff
@@ -240,17 +236,22 @@ class RecordMergingFileIterator(split: HoodieMergeOnReadFileSplit,
   private def merge(curRow: InternalRow, newRecord: HoodieRecord[_]): Option[InternalRow] = {
     // NOTE: We have to pass in Avro Schema used to read from Delta Log file since we invoke combining API
     //       on the record from the Delta Log
+    val curRecord = recordMerger.getRecordType match {
+      case HoodieRecordType.SPARK =>
+        new HoodieSparkRecord(curRow, baseFileReader.schema)
+      case _ =>
+        new HoodieAvroIndexedRecord(serialize(curRow))
+    }
     recordMerger.getRecordType match {
       case HoodieRecordType.SPARK =>
-        val curRecord = new HoodieSparkRecord(curRow, baseFileReader.schema)
-        toScalaOption(recordMerger.merge(curRecord, newRecord, logFileReaderAvroSchema, payloadProps))
+        toScalaOption(recordMerger.merge(curRecord, baseFileReaderAvroSchema, newRecord, logFileReaderAvroSchema, payloadProps))
           .map(r => {
-            val projection = HoodieInternalRowUtils.getCachedUnsafeProjection(r.asInstanceOf[HoodieSparkRecord].getStructType, structTypeSchema)
+            val schema = if (r == curRecord) baseFileReader.schema else logFileReaderStructType
```

Comment:
Please check my comment above regarding the need to return schema form the merger